secrets.go

```diff
@@ -0,0 +1,181 @@
+					envVar, ok := envVarVal.(string)
+					if !ok {
+						return fmt.Errorf("unmarshaling secret: environment_variable must be a string, but was %T", envVarVal)
+					}
+					secret.EnvironmentVariable = envVar
+				}
+
+				*s = append(*s, secret)
+
+			default:
+				return fmt.Errorf("unmarshaling secrets: secret type %T, want string, map[string]any, or *ordered.Map", c)
+			}
+		}
+
+	default:
+		return fmt.Errorf("unmarshaling secrets: got %T, want []any or map[string]any", o)
+	}
+
+	return nil
+}
+
+// MergeWith merges these secrets with another set of secrets, with the other secrets taking precedence.
+// Deduplication is performed based on the EnvironmentVariable field.
+func (s Secrets) MergeWith(other Secrets) Secrets {
+	if len(s) == 0 {
+		return other
+	}
+	if len(other) == 0 {
+		return s
+	}
+
+	// Create a map to track environment variables we've seen for deduplication
+	seen := make(map[string]bool)
+	var result Secrets
+
+	for _, secret := range other {
+		if secret.EnvironmentVariable != "" && !seen[secret.EnvironmentVariable] {
+			result = append(result, secret)
+			seen[secret.EnvironmentVariable] = true
+		}
+	}
+
+	for _, secret := range s {
+		if secret.EnvironmentVariable != "" && !seen[secret.EnvironmentVariable] {
+			result = append(result, secret)
+			seen[secret.EnvironmentVariable] = true
+		}
+	}
+
+	return result
+}
+
+// UnmarshalJSON is used for JSON unmarshaling.
+func (s *Secrets) UnmarshalJSON(b []byte) error {
+	// JSON is just a specific kind of YAML.
+	var n yaml.Node
+	if err := yaml.Unmarshal(b, &n); err != nil {
+		return err
+	}
+	return ordered.Unmarshal(&n, &s)
+}
+
+// MarshalYAML returns the most appropriate YAML representation for the secrets.
+// It prioritizes simple string format, then map format, then returns an error for invalid configurations.
+func (s Secrets) MarshalYAML() (interface{}, error) {
+	if len(s) == 0 {
+		return nil, nil
+	}
+
+	// Use an array of strings if all secrets can be represented as simple strings (key equals environment variable)
+	if canMarshalAsSimpleStrings(s) {
+		result := make([]string, 0, len(s))
+		for _, secret := range s {
+			result = append(result, secret.Key)
+		}
+		return result, nil
+	}
+
+	// Otherwise, use { ENV_VAR:SECRET_NAME } map format if secrets have no other fields
+	if canMarshalAsMap(s) {
+		result := make(map[string]string, len(s))
+		for _, secret := range s {
+			result[secret.EnvironmentVariable] = secret.Key
+		}
+		return result, nil
+	}
+
+	// Cannot represent secrets in user-facing YAML format.
+	// This happens when secrets have RemainingFields or empty EnvironmentVariable.
+	return nil, fmt.Errorf("cannot marshal secrets to YAML: contains secrets with unsupported fields or empty environment variables")
+}
+
+func canMarshalAsSimpleStrings(secrets Secrets) bool {
+	for _, secret := range secrets {
+		if secret.EnvironmentVariable == "" ||
+			secret.Key != secret.EnvironmentVariable ||
+			len(secret.RemainingFields) > 0 {
```

Comment:
If `secret.Key` cannot be empty, then we don't have to separately check if `secret.EnvironmentVariable` is empty, right?

Either way, we're generally not afraid of longer lines in Go. Some people likely disagree, but I think this isn't too long:


```suggestion
		if secret.EnvironmentVariable == "" || secret.Key != secret.EnvironmentVariable || len(secret.RemainingFields) > 0 {
```